consensus/event_loop.go

```diff
@@ -393,14 +395,125 @@ func (fc *FastHotStuffEventLoop) resetScheduledTasks() {
 
 	// Schedule the next block construction task. This will run with currentView param.
-	fc.nextBlockConstructionTask.Schedule(fc.blockConstructionInterval, fc.currentView, fc.onBlockConstructionScheduledTask)
+	fc.nextBlockConstructionTask.Schedule(fc.blockConstructionInterval, fc.currentView, fc.onBlockConstructionScheduledTaskExecuted)
 
 	// Schedule the next timeout task. This will run with currentView param.
 	fc.nextTimeoutTask.Schedule(timeoutDuration, fc.currentView, fc.onTimeoutScheduledTaskExecuted)
 }
 
-func (fc *FastHotStuffEventLoop) onBlockConstructionScheduledTask(blockConstructionView uint64) {
-	// TODO
+// When this function is triggered, it means that we have reached out the block construction
+// time ETA for blockConstructionView. If we have a QC or timeout QC for the view, then we
+// signal the server.
+func (fc *FastHotStuffEventLoop) onBlockConstructionScheduledTaskExecuted(blockConstructionView uint64) {
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Check if the consensus instance is running. If it's not running, then there's nothing
+	// to do here.
+	if fc.status != consensusStatusRunning {
+		return
+	}
+
+	// Check for race conditions where the view advanced at the exact moment this task began
+	// or we have already signaled for this view. If so, then there's nothing to do here.
+	if fc.currentView != blockConstructionView {
+		return
+	}
+
+	// Check if the conditions are met to construct a QC from votes for the chain tip. If so,
+	// we send a signal to the server and cancel all scheduled tasks.
+	if canConstructQC, signersList, aggregatedSignature := fc.tryConstructQCInCurrentView(); canConstructQC {
+		// Signal the server that we can construct a QC for the chain tip
+		fc.ConsensusEvents <- &ConsensusEvent{
+			EventType:   ConsensusEventTypeConstructVoteQC, // The event type
+			View:        fc.currentView,                    // The view that we can construct a block on
+			BlockHash:   fc.getTipBlock().GetBlockHash(),   // The block hash that we can build the QC for
+			BlockHeight: fc.getTipBlock().GetHeight() + 1,  // The block height that we can build the QC for
+			QC: &quorumCertificate{
+				blockHash:           fc.getTipBlock().GetBlockHash(), // The previous block's block hash
+				view:                fc.getTipBlock().GetView(),      // The previous block's view
+				signersList:         signersList,                     // The signers list who voted on the previous block
+				aggregatedSignature: aggregatedSignature,             // Aggregated signature from votes on the previous block
+			},
+		}
+
+		// Cancel the block construction task since we know we can construct a QC in the current view.
+		// It will be rescheduled when we advance view.
+		fc.nextBlockConstructionTask.Cancel()
+
+		return
+	}
+
+	// TODO: Check if we can construct a timeout QC for the current view, and send the signal to the
+	// server
+
+	return
+}
+
+func (fc *FastHotStuffEventLoop) tryConstructQCInCurrentView() (
```

Comment:
Let's add a comment that the caller of this function should hold the consensus lock when calling this function as this function does not hold it itself. also a little more comments on the top of the function providing a little overview would be helpful. Also, should we name this `tryConstructVoteQCInCurrentView`? or will this function also be used for timeout qcs?